### PR TITLE
ClusterAdmin Support 

### PR DESCRIPTION
**Pull request by chandradeepak**
@eapache @buyology @Mongey @mimaison @ongardie-ebay 

I started working on ClusterAdmin support and wrote an initial draft version. Want to get your comments that I am moving in the right direction. After your comments i will close to the pull request.

### CODE REVIEW

File: admin.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+// NewClusterAdmin creates a new ClusterAdmin using the given broker addresses and configuration.
+func NewClusterAdmin(addrs []string, conf *Config) (ClusterAdmin, error) {
+	client, err := NewClient(addrs, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check that we are not dealing with a closed Client before processing any other arguments
+	if client.Closed() {
```
Comment (eapache) [inline admin.go]: unnecessary since you literally just created it

File: admin.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+	ca := &clusterAdmin{
+		client: client,
+		conf:   client.Config(),
+	}
+
+	return ca, nil
+}
+
+func (ca *clusterAdmin) handleResponses(rsp interface{}) {
```
Comment (eapache) [inline admin.go]: I don't understand the motivation for grouping all the response-handling in a giant switch here

File: client.go [hunk 3]
```diff
@@ -35,6 +35,8 @@ type Client interface {
 	// topic/partition, as determined by querying the cluster metadata.
 	Leader(topic string, partitionID int32) (*Broker, error)
 
+	Any() *Broker
```
Comment (eapache) [inline client.go]: I don't think there's a good reason to expose this publicly, it's never been a well-defined method in the first place.

File: admin.go (hunk 1 above)
Comment (chandradeepak) [reply admin.go]: Will delete that

File: admin.go (hunk 2 above)
Comment (chandradeepak) [reply admin.go]: I will handle all the responses . For now I clubbed it but I will handle them individually.

File: client.go (hunk 3 above)
Comment (chandradeepak) [reply client.go]: I need to call broker.Createtopic, broker.Deletetopic etc from admin.go . But I need a broker that is open so I thought of exposing this and accessing a broker which is open. Is there a way to access a broker which is open easily or we need to do that in this file. Which do you recommend ?

Comment (chandradeepak) [reply client.go]: Will use client.Brokers instead of any.

**Comment by chandradeepak** [comment]
@eapache  what is the convenient way to access broker (CreateTopic , DeleteTopic from client. I exposed Any() from client interface but i agree it is not needed.I tried using brokers but it is not guaranteed they are open always. My tests failed with broker not connected.  Can you please suggest any other way to access the underlying brokers which are opened .  I will make  those changes .

**Comment by buyology** [comment]
I would suggest that we first settle on the design in #1048 before finalising the implementation. 

Also, the answer to the question about how to access the broker is that some of the methods (`CreateTopics`, `DeleteTopics`, `CreatePartitions`) must go through the `controller` — and the method for retrieving the controller is not yet implemented.

**Comment by chandradeepak** [comment]
Thanks @buyology  , will wait for the design to be finalized and method for retrieving the controller.

**Comment by chandradeepak** [comment]
@eapache  @buyology , I integrated the controller for this pull request. let me know if you have any comments or changes that needs to be done .

File: admin.go [hunk 4]
```diff
@@ -0,0 +1,360 @@
+package sarama
+
+import "errors"
+
+// ClusterAdmin is the administrative client for Kafka, which supports managing and inspecting topics,
+// brokers, configurations and ACLs. The minimum broker version required is 0.10.0.0.
+// Methods with stricter requirements will specify the minimum broker version required.
+// This client was introduced in 0.11.0.0 and the API is still evolving.
```
Comment (eapache) [inline admin.go]: I don't think this line is necessary in the documentation?

File: admin.go [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+// You MUST call Close() on a client to avoid leaks
```
Comment (eapache) [inline admin.go]: s/client/cluster admin

File: admin.go [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+type ClusterAdmin interface {
+	// Creates a new topic .This operation is supported by brokers with version 0.10.1.0 or higher.
```
Comment (eapache) [inline admin.go]: typo on the period

Comment (eapache) [inline admin.go]: You don't need to specify the version requirement here, I prefer it at the end of the description where it already is for other methods.

File: admin.go [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+	// It may take several seconds after CreateTopic returns success for all the brokers
+	// to become aware that the topics have been created. During this time, ClusterAdmin#listTopics
+	// and ClusterAdmin#describeTopics may not return information about the new topics.
+	// This operation is supported by brokers with version 0.10.1.0 or higher.
+	// The validateOnly option is supported from version 0.10.2.0.
+	CreateTopic(topic string, detail *TopicDetail) error
+
+	// Delete a batch of topics.
```
Comment (eapache) [inline admin.go]: the method only takes one topic

File: admin.go [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+// Methods with stricter requirements will specify the minimum broker version required.
+// This client was introduced in 0.11.0.0 and the API is still evolving.
+// You MUST call Close() on a client to avoid leaks
+type ClusterAdmin interface {
+	// Creates a new topic .This operation is supported by brokers with version 0.10.1.0 or higher.
+	// It may take several seconds after CreateTopic returns success for all the brokers
+	// to become aware that the topics have been created. During this time, ClusterAdmin#listTopics
+	// and ClusterAdmin#describeTopics may not return information about the new topics.
+	// This operation is supported by brokers with version 0.10.1.0 or higher.
+	// The validateOnly option is supported from version 0.10.2.0.
```
Comment (eapache) [inline admin.go]: the method doesn't take a `validateOnly` flag

File: admin.go [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+	// The resources with their configs (topic is the only resource type with configs
+	// that can be updated currently Updates are not transactional so they may succeed
+	// for some resources while fail for others. The configs for
+	// a particular resource are updated atomically.
+	AlterConfig(resourceType ConfigResourceType, name string, entries map[string]*string, validateOnly bool) error
+
+	// Creates access control lists (ACLs) which are bound to specific resources.
+	// This operation is not transactional so it may succeed for some ACLs while fail for others.
+	// If you attempt to add an ACL that duplicates an existing ACL, no error will be raised, but
+	// no changes will be made.This operation is supported by brokers with version 0.11.0.0 or higher.
```
Comment (eapache) [inline admin.go]: missing space

File: admin.go [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+	CreateACL(resource Resource, acl Acl) error
+
+	ListAcls(filter AclFilter) ([]ResourceAcls, error)
```
Comment (eapache) [inline admin.go]: missing description

File: admin.go [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+}
+
+func (ca *clusterAdmin) CreateTopic(topic string, detail *TopicDetail) error {
+
+	if topic == "" {
+		return ErrInvalidTopic
+	}
+
+	if detail == nil {
+		return ErrInvalidInput
```
Comment (eapache) [inline admin.go]: I don't think it's worth naming an error for this, it's better to give something less computer-friendly and more human-friendly (e.g. `errors.New("You must specify topic details")`)

File: admin.go (hunk 4 above)
Comment (chandradeepak) [reply admin.go]: ok. will remove that . I was thinking mentioning the version dependency would be needed especially admin related stuff is supported only in latest versions of kafka.

File: admin.go (hunk 6 above)
Comment (chandradeepak) [reply admin.go]: will address that.

Comment (chandradeepak) [reply admin.go]: ok , will do that.

File: admin.go (hunk 7 above)
Comment (chandradeepak) [reply admin.go]: that is what we though in the initial design. So i went with that . Should i change it to take multiple topics ?

File: admin.go (hunk 11 above)
Comment (chandradeepak) [reply admin.go]: ok sure , will do that

File: admin.go (hunk 7 above)
Comment (eapache) [reply admin.go]: Probably? At the very least, the code and the documentation should be consistent.

File: admin.go (hunk 4 above)
Comment (chandradeepak) [reply admin.go]: removed it

File: admin.go (hunk 6 above)
Comment (chandradeepak) [reply admin.go]: done

File: admin.go (hunk 8 above)
Comment (chandradeepak) [reply admin.go]: gave option for  validateOnly flag

File: admin.go (hunk 9 above)
Comment (chandradeepak) [reply admin.go]: done

File: admin.go (hunk 10 above)
Comment (chandradeepak) [reply admin.go]: gave support for ListAcls and added the description for it.

File: admin.go (hunk 11 above)
Comment (chandradeepak) [reply admin.go]: removed the ErrInvalidInput from errors.go

**Comment by eapache** [comment]
Did you intend to push https://github.com/Shopify/sarama/pull/1116 to this PR instead?

File: admin.go [hunk 12]
```diff
@@ -0,0 +1,376 @@
+package sarama
+
+import "errors"
+
+type ClusterAdmin interface {
```
Comment (eapache) [inline admin.go]: Sorry my previous comment wasn't clear. I still think we should have a description for `ClusterAdmin`, I just think it doesn't need to mention the bit about "This client was introduced in 0.11.0.0 and the API is still evolving."

File: admin.go [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+	// Creates a new topic. This operation is supported by brokers with version 0.10.1.0 or higher.
```
Comment (eapache) [inline admin.go]: You've still got a duplicate version sentence here.

File: admin.go [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+	// It may take several seconds after CreateTopic returns success for all the brokers
+	// to become aware that the topic has been created. During this time, ClusterAdmin#listTopics
+	// and ClusterAdmin#describeTopics may not return information about the new topic.
+	// This operation is supported by brokers with version 0.10.1.0 or higher.
+	// The validateOnly option is supported from version 0.10.2.0.
+	CreateTopic(topic string, detail *TopicDetail, validateOnly bool) error
+
+	// Delete a batch of topics.
```
Comment (eapache) [inline admin.go]: This sentence needs to be singular still since the method only deletes one at a time.

File: admin.go [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+
+	// Increase the number of partitions of the topics  according to the corresponding values.
+	// If partitions are increased for a topic that has a key, the partition logic or ordering of
+	// the messages will be affected. It may take several seconds after this method returns
+	// success for all the brokers to become aware that the partitions have been created.
+	// During this time, ClusterAdmin#describeTopics may not return information about the
+	// new partitions. This operation is supported by brokers with version 1.0.0 or higher.
+	CreatePartitions(topic string, count int32, assignment [][]int32, validateOnly bool) error
+
+	//Delete records whose offset is smaller than the given offset of the corresponding partition.
```
Comment (eapache) [inline admin.go]: spacing after the `//`

File: errors.go [hunk 16]
```diff
@@ -49,6 +49,8 @@ var ErrControllerNotAvailable = errors.New("kafka: controller is not avaiable")
 // the metadata.
 var ErrNoTopicsToUpdateMetadata = errors.New("kafka: no specific topics to update metadata")
 
+var ErrReferenceNodeNotFound = errors.New("kafka: response did not contain a reference to node")
```
Comment (eapache) [inline errors.go]: Is this fundamentally any different from `ErrIncompleteResponse`?

Comment (chandradeepak) [reply errors.go]: i was looking at the java documentation in kafka when ever topics creation is not successful or the response we dont get a response to the topic which we requested changes for it was returning some thing like that . I can change it to use ErrIncompleteResponse but it doesn't say the intent that the specified resource is not found.  I will wait for your response and take that action.

Comment (eapache) [reply errors.go]: I'm not sure I fully understand, if something doesn't work the broker should still reply with that topic/partition, just with an error code set? If it's omitting the topic/partition entirely that's new behaviour.

At least as written, NodeNotFound doesn't contain any more information beyond IncompleteResponse about the "why", so I'd prefer to stick with IncompleteResponse until we have something more meaningful to put here.

Comment (chandradeepak) [reply errors.go]: ok sure will switch to do that .

**Comment by eapache** [comment]
This looks pretty good to me now. Could you please squash/rebase your commits?

**Comment by chandradeepak** [comment]
Done squashing all the commits .

**Comment by eapache** [comment]
Thanks!

**Pull request merged**